dev/com.ibm.ws.jpa.container.ormdiagnostics/src/com/ibm/ws/jpa/diagnostics/AbstractIntrospection.java

```diff
@@ -0,0 +1,298 @@
+/*******************************************************************************
+ * Copyright (c) 2018 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.jpa.diagnostics;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class AbstractIntrospection implements PlatformIntrospection {
+    private final static String[] openJPAEntityManagerFactoryImplClasses = { "org.apache.openjpa.persistence.EntityManagerFactoryImpl",
+                                                                             "com.ibm.ws.persistence.EntityManagerFactoryImpl" };
+    private final static String[] eclipselinkEntityManagerFactoryImplClasses = { "org.eclipse.persistence.internal.jpa.EntityManagerFactoryImpl" };
+
+    public static boolean isSupportedPlatform(Class jpaProviderImplClass) {
+        if (jpaProviderImplClass == null) {
+            return false;
+        }
+
+        final String className = jpaProviderImplClass.getName();
+
+        if (eclipselinkEntityManagerFactoryImplClasses[0].equals(className)) {
+            return true;
+        }
+
+        if (openJPAEntityManagerFactoryImplClasses[0].equals(className) || openJPAEntityManagerFactoryImplClasses[1].equals(className)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    public static PlatformIntrospection getPlatformIntrospection(Class jpaProviderImplClass) {
+        if (jpaProviderImplClass == null) {
+            return null;
+        }
+
+        final String className = jpaProviderImplClass.getName();
+
+        if (eclipselinkEntityManagerFactoryImplClasses[0].equals(className)) {
+            return new EclipselinkIntrospection();
+        }
+
+        if (openJPAEntityManagerFactoryImplClasses[0].equals(className) || openJPAEntityManagerFactoryImplClasses[1].equals(className)) {
+            return new OpenJPAIntrospection();
+        }
+
+        return null;
+    }
+
+    /*
+     * Utility methods used for introspecting JPA Persistence Provider Implementation Data Structures
+     */
+    protected static Object reflectObjValue(Object o, String field) throws Exception {
+        final Class<?> c = o.getClass();
+        final Field f = findField(c, field); //        c.getField(field);
+        if (f == null) {
+            return null;
+        }
+        final boolean accessible = f.isAccessible();
```

Comment:
Note: java.lang.Class.getDeclaredField will return a copy of the Field.... so changes made here with setAccessible are only visible here.... and do not need to be undone.

If it is possible to cache this Field, that would be nice, but probably not important if this won't run very often.

Will this run in a doPriv? or an a privileged thread? If running as regular trace, you may fail Java 2 security authorization.

Reply:
I believe the introspection dump threads are privileged threads, I've been running the FATs (which enable java2sec) and the dump has not been failing with ACE involving reflection (only ACE is from the use of the jdk provided JAXB provider impl)